### PR TITLE
[OptApp] Making the control projection modular

### PR DESCRIPTION
**Pull request by sunethwarna**
**📝 Description**
This PR makes the projection mechanisms used in the `Control`s modular such that, user can specify which type of projection to be used when updating physical fields, and physical field gradients.

Three type of projection methods are implemented:
1. `IdentityProjection` -> y = x (no projection)
2. `SigmoidalProjection` -> y = h(x) (based on modified sigmoidal function)
3. `AdaptiveSigmoidalProjection` -> y = g(x, step) (based on modified sigmoidal function where `beta` is changed with optimization iterations.

**Regarding the ref value changes**

Upto now, the `shell_thickness_control` only updated the sigmoidal projection beta value if a valid update is carried out, which is not the correct behaviour. This PR changes it to always update beta irrespective of whether a valid update is carried out. Hence, the ref values are changed. This behaviour is desired because, even in the case where optimizer converges to an intermeidate state with the local minimum, this will push the values to goto the boundaries, which is important for discrete value optimization.

**🆕 Changelog**
- Added `Projection` base class
- Added `IdentityProjection`
- Added `SigmoidalProjection`
- Added `AdaptiveSigmoidalProjection`
- Updated the Controls accordingly
- Updated the tests accordingly.

### CODE REVIEW

**Review by Igarizza — COMMENTED** [review]
@sunethwarna Thank you for your work! Can we apply previously used Projection to the RGP test to keep the ref untouched?

**Comment by sunethwarna** [comment]
@Igarizza Ok, I have moved the ref value changes to a seperate PR so it can be safely reviewed in #12707. Could you take a look? After merging that PR, I will update this PR.

File: applications/OptimizationApplication/python_scripts/utilities/opt_projection.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
+from typing import Optional
+from abc import ABC, abstractmethod
+
+import KratosMultiphysics as Kratos
+import KratosMultiphysics.OptimizationApplication as KratosOA
+from KratosMultiphysics.OptimizationApplication.utilities.optimization_problem import OptimizationProblem
+from KratosMultiphysics.OptimizationApplication.utilities.union_utilities import ContainerExpressionTypes
+
+class Projection(ABC):
+    """Projection methods to convert given x space values to given y space values
```
Comment (Igarizza) [inline applications/OptimizationApplication/python_scripts/utilities/opt_projection.py]: Projection is too general name for class name and file. What about DesignVariableProjection or DesignVariableProjectionMethod ???

**Review by Igarizza — COMMENTED** [review]
Super cool! Thnx @sunethwarna I have only small comment to improve naming.

File: applications/OptimizationApplication/python_scripts/utilities/opt_projection.py (hunk 1 above)
Comment (sunethwarna) [reply applications/OptimizationApplication/python_scripts/utilities/opt_projection.py]: done :)

**Review by Igarizza — APPROVED** [review]

**Pull request merged**